submissions/BlueLamantine/friends app/js/main.js

```diff
@@ -0,0 +1,339 @@
+  }
+
+  startFilter() {
+    this.usersByDefault = [...this.usersDataForFilter];
+    this.filtersPanel.addEventListener('click', ({ target }) => {
+      if (target.type === 'radio' || target.type === 'text') {
+        this.selectSort(target);
+        this.renderDataWithFilters.usersDataForRender = this.usersDataForFilter;
+        this.renderDataWithFilters.renderProfiles();
+      }
+    });
+  }
+
+  sortByGender(gender) {
+    this.usersDataForFilter = [...this.usersByDefault];
+    const selectGender = {
+      allGender: () => {
+        this.usersDataForFilter = this.sortedUsers = [...this.usersByDefault];
+      },
+      male: () => {
+        this.usersDataForFilter = this.usersDataForFilter.filter(
+          user => user.gender == 'male'
+        );
+        this.sortedUsers = [...this.usersDataForFilter];
+      },
+      female: () => {
+        this.usersDataForFilter = this.usersDataForFilter.filter(
+          user => user.gender == 'female'
+        );
+        this.sortedUsers = [...this.usersDataForFilter];
+      },
+    };
+    return selectGender[gender]();
+  }
+
+  sortByContent(sortType, sortTypes, sortBy) {
+    if (this.sortedUsers.length !== 0)
+      this.usersDataForFilter = [...this.sortedUsers];
+    else this.sortedUsers = [...this.usersByDefault];
+    const typeOfSort = {
+      [sortTypes.byDefault]: () => {
+        this.usersDataForFilter = [...this.sortedUsers];
+      },
+      [sortTypes.inAscending]: () => {
+        if (sortBy == 'name')
+          this.usersDataForFilter.sort((a, b) =>
+            a.name.first.localeCompare(b.name.first)
+          );
+        if (sortBy == 'age')
+          this.usersDataForFilter.sort((a, b) =>
+            a.dob.age.toString().localeCompare(b.dob.age.toString())
+          );
+      },
+      [sortTypes.inDescending]: () => {
+        if (sortBy == 'name')
+          this.usersDataForFilter.sort((a, b) =>
+            b.name.first.localeCompare(a.name.first)
+          );
+        if (sortBy == 'age')
+          this.usersDataForFilter.sort((a, b) =>
+            b.dob.age.toString().localeCompare(a.dob.age.toString())
+          );
+      },
+    };
+    return typeOfSort[sortType]();
+  }
+}
+
+class Users {
+  constructor(render, filter) {
+    this.renderData = render;
+    this.filterData = filter;
+    this.apiURL = `https://randomuser.me/api/`;
+    this.errorMessage = `Uh oh, something has gone wrong. Please tweet us @randomapi about the issue. Thank you.`;
+  }
+
+  async getUsersData() {
+    try {
+      const response = await fetch(this.apiURL + `?results=100&seed=abc`);
+      if (!response.ok) throw new Error(errorMessage);
+      const usersData = await response.json();
+      this.renderData.usersDataForRender = this.filterData.usersDataForFilter =
+        usersData.results;
+      this.renderData.renderProfiles();
+      this.filterData.startFilter();
+    } catch (err) {
+      console.error(err);
+    }
+  }
+}
+
+class Observer {
```

Comment:
Same, why class and not plain function?

Reply:
I removed this class and move it's only one function to hide/show menu to constructor of class Render.